Mail/Views/Thread List/DragAndDropModifier.swift

```diff
@@ -0,0 +1,124 @@
+    @Entry
+    var isHovered = false
+}
+
+struct DropThreadViewModifier: ViewModifier {
+    @EnvironmentObject private var mailboxManager: MailboxManager
+    @EnvironmentObject private var actionsManager: ActionsManager
+
+    @State private var isDropTargeted = false
+
+    let destinationFolder: Folder
+
+    func body(content: Content) -> some View {
+        if #available(macCatalyst 16.0, iOS 16.0, *) {
+            content
+                .contentShape(Rectangle())
+                .dropDestination(for: DraggedThread.self) { draggedThreads, _ in
+                    let flatThreads = draggedThreads.flatMap(\.threadIds)
+                    handleDroppedThreads(flatThreads)
+
+                    return true
+                } isTargeted: {
+                    isDropTargeted = $0
+                }
+                .environment(\.isHovered, isDropTargeted)
+        } else {
+            content
+        }
+    }
+
+    private func handleDroppedThreads(_ draggedThreadsIds: [String]) {
+        let threads = draggedThreadsIds.compactMap {
+            mailboxManager.getThread(from: $0)
+        }
+
+        guard let originFolder = threads.first(where: { $0.folder != nil })?.folder,
+              originFolder != destinationFolder else {
+            return
+        }
+
+        let movedMessages = threads.flatMap(\.messages)
+
+        Task {
+            await tryOrDisplayError {
+                try await actionsManager.performMove(
+                    messages: movedMessages,
+                    from: originFolder,
+                    to: destinationFolder
+                )
+            }
+        }
+    }
+}
+
+struct DraggableThreadViewModifier: ViewModifier {
+    @AppStorage(UserDefaults.shared.key(.accentColor)) private var accentColor = DefaultPreferences.accentColor
+    let draggedThreadId: [String]
+
+    func body(content: Content) -> some View {
+        if #available(macCatalyst 16.0, iOS 16.0, *) {
+            content
+            #if os(macOS) || targetEnvironment(macCatalyst)
+            .draggable(DraggedThread(threadIds: draggedThreadId)) {
+                ZStack(alignment: .bottomTrailing) {
+                    Image(systemName: "envelope.fill")
+                        .font(MailTextStyle.labelDraggableThread.font)
+                        .padding(value: .small)
+                    if draggedThreadId.count > 1 {
+                        Text("\(draggedThreadId.count)")
+                            .padding(value: .small)
+                            .background(accentColor.secondary.swiftUIColor, in: .circle)
+                            .font(MailTextStyle.bodySmallAccent.font)
+                    }
+                }
+            }
+            #else
+            .draggable(DraggedThread(threadIds: draggedThreadId))
+            #endif
+        } else {
+            content
+        }
+    }
+}
+
+extension View {
+    func dropThreadHandler(destinationFolder: Folder) -> some View {
```

Comment:
If the second modifier is called `draggableThread`, could we name this one `droppableThread` ?
@PhilippeWeidmann

Reply:
Indeed this would make more sense !